src/common/components/ui/formik/validationSchemas.ts

```diff
@@ -466,3 +475,63 @@ export const locationValidationSchema = Yup.string()
     message: LOCATION_VALIDATION_MESSAGES.INVALID_VALUE,
     excludeEmptyString: true,
   });
+
+export const machineNetworksValidationSchema = Yup.array().of(
+  Yup.object({ cidr: ipBlockValidationSchema, clusterId: Yup.string() }),
+);
+
+export const clusterNetworksValidationSchema = Yup.array().of(
+  Yup.lazy((values: ClusterNetwork) =>
+    Yup.object({
+      cidr: ipBlockValidationSchema,
+      hostPrefix: hostPrefixValidationSchema({ clusterNetworkCidr: values.cidr }),
+      clusterId: Yup.string(),
+    }),
+  ),
+);
+
+export const serviceNetworkValidationSchema = Yup.array().of(
+  Yup.object({
+    cidr: ipBlockValidationSchema,
+    clusterId: Yup.string(),
+  }),
+);
+
+export const dualStackValidationSchema = (field: string) =>
+  Yup.array()
+    .max(2, `Maximum number of ${field} subnets in dual stack is 2.`)
+    .test(
+      'dual-stack-ipv4',
+      'First network has to be IPv4 subnet.',
+      (values) => values[0].cidr && Address4.isValid(values[0].cidr),
+    )
+    .test(
+      'dual-stack-ipv6',
+      'Second network has to be IPv6 subnet.',
+      (values) => values[1].cidr && Address6.isValid(values[1].cidr),
+    );
+
+export const isSingleStack = (
+  networks1?: (MachineNetwork | ClusterNetwork | ServiceNetwork)[],
+  networks2?: (MachineNetwork | ClusterNetwork | ServiceNetwork)[],
+  networks3?: (MachineNetwork | ClusterNetwork | ServiceNetwork)[],
+) =>
+  !!(
+    (networks1?.every((network) => network.cidr && Address4.isValid(network.cidr)) &&
```

Comment:
I would extract this logic to a reusable function, as I see the same code is used later in a Component as well.

So you can define either one function and pass a parameter to check for V4 or V6, or create two separate functions for that. (I couldn't think of a better name for the function,)

```
const networkOnlyHasIP4Addresses = (networkList)  => ...;
const networkOnlyHasIP6Addresses = (networkList)  => ...;
```

Then the main code is easier to read:

```
const hasAllIPv4 = networkOnlyHasIP4Addresses(network1) && ...networkOnlyHasIP4Addresses(network3);

const hasAllIPv6 = networkOnlyHasIP6Addresses(network1) && ...networkOnlyHasIP6Addressesd(network3);

return hasAllIPv4 || hasAllIPv6;
```


Also, if the logic about validating the IPs is not straightforward, I think an early return is better.
Meaning, if you get `true` on allV4, you can return true and avoid evaluating all the ipV6 calls.
```
const hasAllIPv4 = networkOnlyHasIP4Addresses(network1) && ...networkOnlyHasIP4Addresses(network3);
if (hasAllIPv4) {
  return true;
}
const hasAllIPv6 = networkOnlyHasIP6Addresses(network1) && ...networkOnlyHasIP6Addressesd(network3);

return hasAllIPv6;
```

Reply:
yeah, I agree, it would be nice to have this logic in a function

Reply:
I have added two functions: `allSubnetsIPv4`and `allSubnetsIPv6` functions and I'm using those.